ufl/argument.py

```diff
@@ -144,6 +142,70 @@ def __eq__(self, other):
                 self._part == other._part and
                 self._ufl_function_space == other._ufl_function_space)
 
+
+@ufl_type()
+class Argument(FormArgument, BaseArgument):
+    """UFL value: Representation of an argument to a form."""
+    __slots__ = (
+        "_ufl_function_space",
+        "_ufl_shape",
+        "_number",
+        "_part",
+        "_repr",
+    )
+
+    _primal = True
+    _dual = False
+
+    __getnewargs__ = BaseArgument.__getnewargs__
+    __str__ = BaseArgument.__str__
+    _ufl_signature_data_ = BaseArgument._ufl_signature_data_
+
+    def __new__(cls, *args, **kw):
+        if args[0] and is_dual(args[0]):
+            return Coargument(*args, **kw)
+        return super().__new__(cls)
+
+    def __init__(self, function_space, number, part=None):
+        FormArgument.__init__(self)
+        BaseArgument.__init__(self, function_space, number, part)
+
+    def ufl_domains(self):
+        return BaseArgument.ufl_domains(self)
+
+    def __repr__(self):
+        return self._repr
+
+
+class Coargument(BaseForm, BaseArgument):
```

Comment:
At the moment it is possible to write `Argument(V.dual())` and it will return a cofunction object. Since objects in the dual space behave like forms they need to inherit from BaseForm which we can't do if they are the same class as the primal space objects.

Reply:
Ok I see, it makes sense.